gen_breaktest.go

```diff
@@ -0,0 +1,215 @@
+		}
+		original, expected, err := parseRuneSequence(line, original[:0], expected[:0])
+		if err != nil {
+			return nil, fmt.Errorf(`line %d: %v: %q`, num, err, line)
+		}
+		fmt.Fprintf(buf, "\t{original: \"%s\", expected: %s}, // %s\n", original, expected, comment)
+	}
+	if err := sc.Err(); err != nil {
+		return nil, err
+	}
+
+	// Check for final "# EOF", useful check if we're streaming via HTTP
+	if !bytes.Equal(line, []byte("# EOF")) {
+		return nil, fmt.Errorf(`line %d: exected "# EOF" as final line, got %q`, num, line)
+	}
+	buf.WriteString("}\n")
+	log.Printf("processed %d lines to %d/%d bytes", num, buf.Len(), buf.Cap())
+	return buf.Bytes(), nil
+}
+
+// used by parseRuneSequence to match input via bytes.HasPrefix
+var (
+	prefix  = []byte("÷ ")
+	breakOk = []byte("÷")
+	breakNo = []byte("×")
+)
+
+// parseRuneSequence parses a rune + breaking opportunity sequence from b
+// and appends the Go code for testcase.original to orig
+// and appends the Go code for testcase.expected to exp.
+// It retuns the new orig and exp slices.
+//
+// E.g. for the input b="÷ 0020 × 0308 ÷ 1F1E6 ÷"
+// it will append
+//     "\u0020\u0308\U0001F1E6"
+// and "[][]rune{{0x0020,0x0308},{0x1F1E6},}"
+// to orig and exp respectively.
+//
+// The formatting of exp is expected to be cleaned up by gofmt or format.Source.
+// Note we explicitly require the sequence to start with ÷ and we implicitly
+// require it to end with ÷.
+func parseRuneSequence(b, orig, exp []byte) ([]byte, []byte, error) {
+	// Check for and remove first ÷.
+	if !bytes.HasPrefix(b, prefix) {
+		return nil, nil, fmt.Errorf("expected line to start with %q", prefix)
+	}
+	b = b[len(prefix):]
+
+	boundary := true
+	exp = append(exp, "[][]rune{"...)
+	for len(b) > 0 {
+		if boundary {
+			exp = append(exp, '{')
+		}
+		exp = append(exp, "0x"...)
+		// Find end of hex digits.
+		var i int
+		for i = 0; i < len(b) && b[i] != ' '; i++ {
+			if d := b[i]; ('0' <= d || d <= '9') ||
+				('A' <= d || d <= 'F') ||
+				('a' <= d || d <= 'f') {
+				continue
+			}
+			return nil, nil, errors.New("bad hex digit")
+		}
+		switch i {
+		case 4:
+			orig = append(orig, "\\u"...)
+		case 5:
+			orig = append(orig, "\\U000"...)
+		default:
+			return nil, nil, errors.New("unsupport code point hex length")
+		}
+		orig = append(orig, b[:i]...)
+		exp = append(exp, b[:i]...)
+		b = b[i:]
+
+		// Check for space between hex and ÷ or ×.
+		if len(b) < 1 || b[0] != ' ' {
+			return nil, nil, errors.New("bad input")
+		}
+		b = b[1:]
+
+		// Check for next boundary.
+		switch {
+		case bytes.HasPrefix(b, breakOk):
+			boundary = true
+			b = b[len(breakOk):]
+		case bytes.HasPrefix(b, breakNo):
+			boundary = false
+			b = b[len(breakNo):]
+		default:
+			return nil, nil, errors.New("missing ÷ or ×")
+		}
+		if boundary {
+			exp = append(exp, '}')
+		}
+		exp = append(exp, ',')
+		if len(b) > 0 && b[0] == ' ' {
+			b = b[1:]
+		}
+	}
+	exp = append(exp, '}')
+	return orig, exp, nil
+}
```

Comment:
Just for fun, I tried this with Unicode version 11.0.0 and got this message:

```
gen_breaktest: using "https://www.unicode.org/Public/11.0.0/ucd/auxiliary/GraphemeBreakTest.txt"
gen_breaktest: processed 700 lines to 128064/245962 bytes
gen_breaktest: gofmt: 43:21: escape sequence is invalid Unicode code point (and 71 more errors)
exit status 1
gen_breaktest.go:19: running "go": exit status 1
```

I didn't check any further what changed since then and it's not important now anyway. But we should expect breaking changes in the official test case file. I would like you to be available for adjustments to your code in case it fails some time in the future. Please let me know if that's ok with you.

Reply:
Interesting. By the way, that's why the `//srcfmt = src` is there after the `log.Fatalln` in `main`, so that by moving the comment the un-formatted Go source can be generated as-is and examined by hand if/when required. In this case it seems that `"\uD800"` is not valid Go (it seems it's because it's a surrogate, I don't know Unicode that well). It seems the only changes from `GraphemeBreakTest-11.0.0.txt` to `GraphemeBreakTest-12.0.0.txt` were to completely remove all test cases that had D800 in them. :slightly_smiling_face:

> I would like you to be available for adjustments to your code

Sure, I should be reachable via Github notification/tagging for the foreseeable future. Of course this would only happen if (a) they changed the format of the file or some other breaking change and (b) someone wanted to implement the next Unicode version and manually bumped the version number to try and get new testcases. As with the current situation, the existing test cases are still there and still work. So hopeful any such breakage would be (a) unlikely and (b) not urgent.

Reply:
Yeah, I would think so, too. I mean, from what I can see, this PR is written quite well. I don't expect any problems with it. But it is more than just a minor change, and in those cases, I'd like to be able to get back to you in cases something breaks in these parts.